### PR TITLE
[Experimental]: Improve variation selector frontend

### PR DESCRIPTION
**Pull request by octaedro**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Partially addresses [53012](https://github.com/woocommerce/woocommerce/issues/53012).

This PR adds the code to show the variation selector on the page frontend. The change is part of [a bigger PR](https://github.com/woocommerce/woocommerce/pull/53855) split to make reviewing easier.

Screenshot 2024-12-23 at 10 44 30 AM

The selector is not tied yet with the `Add to cart` button, so only test the variation selection and `clear` button.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

Having `WooCommerce Beta Tester` installed and activater
1. Go to Tools > WCA Test Helper > Features
2. Enable the features: `blockified-add-to-cart` and `experimental-blocks`
3. Create a variable product with variations.
4. Go to Pages > Add New Page
5. Add a `Single Product` block and select the product you created in step 3.
6. Add an `Add to Cart with Option (Experimental)` block.
7. Verify that the selectors with the attributes are visible.
8. Save the page and view it.
9. There should be one selector per attribute you choose for your variations.
10. Pick one, and a `Clear` button should appear.
11. Press it to clear the selectors.

https://github.com/user-attachments/assets/b8369d1a-7d54-415a-aa3f-8b61a3288627

12. Go to the variable product you created in step 3. Go to Product data > Variations and select a default variation in `Default Form Values`.
13. Verify that the variation you pick to be shown by default is shown correctly.

Screenshot 2024-12-23 at 11 27 41 AM

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/src/Blocks/BlockTypes/AddToCartWithOptionsVariationSelector.php [hunk 1]
```diff
@@ -17,16 +17,348 @@ class AddToCartWithOptionsVariationSelector extends AbstractBlock {
...
+	 * @param array      $attributes Block attributes.
+	 * @return string Rendered form HTML
+	 */
+	private function render_variation_form( $product, $attributes ): string {
+		$variation_attributes = $product->get_variation_attributes();
+		if ( empty( $variation_attributes ) ) {
+			return '';
+		}
+
+		wp_enqueue_script( 'wc-add-to-cart-variation' );
```
Comment (octaedro) [inline plugins/woocommerce/src/Blocks/BlockTypes/AddToCartWithOptionsVariationSelector.php]: We should rewrite [this file](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/client/legacy/js/frontend/add-to-cart-variation.js). I'll create an issue to tackle that. Until then, we use the same code that we use for variable products on the product page.

File: plugins/woocommerce/src/Blocks/BlockTypes/AddToCartWithOptionsVariationSelector.php [hunk 2]
```diff
@@ -17,16 +17,313 @@ class AddToCartWithOptionsVariationSelector extends AbstractBlock {
...
```
Comment (Aljullu) [inline plugins/woocommerce/src/Blocks/BlockTypes/AddToCartWithOptionsVariationSelector.php]: Can we move this line under

```PHP
		$variations = $this->get_variations_data( $product );
		if ( empty( $variations ) ) {
			return '';
		}
```

No need to enqueue the script if we have no variations, no?

File: plugins/woocommerce/src/Blocks/BlockTypes/AddToCartWithOptionsVariationSelector.php [hunk 3]
```diff
@@ -17,16 +17,313 @@ class AddToCartWithOptionsVariationSelector extends AbstractBlock {
...
+	 * @param WC_Product $product Product instance.
+	 * @param array      $variations Variations data.
+	 * @param array      $variation_attributes Variation attributes.
+	 * @return string Form HTML
+	 */
+	private function get_form_html( $product, $variations, $variation_attributes ): string {
+		$variations_json = wp_json_encode( $variations );
+		$variations_attr = function_exists( 'wc_esc_json' ) ? wc_esc_json( $variations_json ) : _wp_specialchars( $variations_json, ENT_QUOTES, 'UTF-8', true );
+
+		$wrapper_attributes = get_block_wrapper_attributes( array( 'class' => 'wc-block-add-to-cart-form' ) );
```
Comment (Aljullu) [inline plugins/woocommerce/src/Blocks/BlockTypes/AddToCartWithOptionsVariationSelector.php]: Why do we need the wrapper `<div>`? Won't this block be inside the Add to Cart with Options block anyway? :thinking:

File: plugins/woocommerce/src/Blocks/BlockTypes/AddToCartWithOptionsVariationSelector.php [hunk 4]
```diff
@@ -17,16 +17,313 @@ class AddToCartWithOptionsVariationSelector extends AbstractBlock {
...
+	protected function render_variation_selector( $product, $attribute_name, $options ): string {
+		$selected  = $this->get_selected_attribute_value( $product, $attribute_name );
+		$select_id = esc_attr( 'attribute_' . sanitize_title( $attribute_name ) );
+
+		$html = sprintf(
+			'<select id="%1$s"
+				class="wc-block-components-product-add-to-cart-attribute-select"
+				name="%1$s"
+				data-attribute_name="attribute_%2$s"
+				data-wc-on--change="actions.updateVariation">',
```
Comment (Aljullu) [inline plugins/woocommerce/src/Blocks/BlockTypes/AddToCartWithOptionsVariationSelector.php]: Currently `actions.updateVariation` is not hooked to any action, no? Wondering if we should remove it for now or if there is a reason to keep it.

**Review by Aljullu — APPROVED** [review]
LGTM, thanks for splitting this into two different PRs, making it easier to review!

As you mentioned, we will probably need to make some updates to make it work with the iAPI instead of the legacy script. But for now, I think it's good to go. :100:

File: plugins/woocommerce/src/Blocks/BlockTypes/AddToCartWithOptionsVariationSelector.php (hunk 3 above)
Comment (octaedro) [reply plugins/woocommerce/src/Blocks/BlockTypes/AddToCartWithOptionsVariationSelector.php]: Good call! I removed it in the commit [b3097c1](https://github.com/woocommerce/woocommerce/pull/53947/commits/b3097c15757e5c4227d09ce34c60751b2b002ed9)

File: plugins/woocommerce/src/Blocks/BlockTypes/AddToCartWithOptionsVariationSelector.php (hunk 4 above)
Comment (octaedro) [reply plugins/woocommerce/src/Blocks/BlockTypes/AddToCartWithOptionsVariationSelector.php]: Good idea! [cda54c1](https://github.com/woocommerce/woocommerce/pull/53947/commits/cda54c1e1f316cddaf7f4eb165462684f4208873)

File: plugins/woocommerce/src/Blocks/BlockTypes/AddToCartWithOptionsVariationSelector.php (hunk 2 above)
Comment (octaedro) [reply plugins/woocommerce/src/Blocks/BlockTypes/AddToCartWithOptionsVariationSelector.php]: Done in [3c97fb4](https://github.com/woocommerce/woocommerce/pull/53947/commits/3c97fb48a3fe26809fb934151c2154fa4e3259e7).

**Pull request merged**